scrapinghub/client/utils.py

```diff
@@ -136,40 +137,31 @@ class _MappingProxy(_ProxyBase):
 
     def __init__(self, cls, client, key):
         super(_MappingProxy, self).__init__(cls, client, key)
-        methods = ['ignore_fields', 'expire', 'save', 'liveget']
-        self._proxy_methods(methods)
-        self.iterkeys = self.__iter__
-
-    def __str__(self):
-        return str(self._origin._data)
-
-    def __repr__(self):
-        return '{}({})'.format(self.__class__.__name__,
-                               repr(self._origin._data))
-
-    def __getitem__(self, key):
-        return self._origin[key]
-
-    def __setitem__(self, key, value):
-        self._origin[key] = value
 
-    def __delitem__(self, key):
-        del self._origin[key]
+        # cls_methods methods will be imported from origin class
+        origin_cls_methods = ['__iter__', '__len__', '__str__', '__repr__',
+                              '__getitem__', '__setitem__', '__delitem__']
 
-    def __iter__(self):
-        return iter(self._origin)
+        other_attrs = ['ignore_fields', 'expire', '_data', '_deleted',
+                       'save', 'liveget', ('iter', '__iter__')]
 
-    def __len__(self):
-        return len(self._origin)
+        # blacklisted attributes should be just skipped on proxying
+        blacklisted_attrs = ['__class__', '__init__', '__str__', '__repr__']
 
-    def keys(self):
-        return list(self.iterkeys())
+        # proxy all the main methods from MutableMapping
+        methods = [method for method in dir(MutableMapping)
+                   if callable(getattr(MutableMapping, method))
+                   and method not in origin_cls_methods
+                   and method not in blacklisted_attrs]
+        proxy_methods(MutableMapping, self.__class__, methods, force=True)
```

Comment:
@chekunkov 
> I don't follow why special methods has to be proxied at the class level...
http://stackoverflow.com/questions/10484304/setting-special-methods-using-setattr
https://paste.scrapinghub.com/show/11632/
So in Python 3 it works via rewriting methods at the class level, and in Python 2 it doesn't work at all, what I'm missing here?

> does this list include methods that should be proxied with a force option?
It does, empty class object already contains a lot of the methods from that list:
```
In [9]: class X():
   ...:     pass
In [10]: dir(X())
Out[10]:
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__']
```